File: cpp/0198_House_Robber.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <queue>
#include <stack>
#include <deque>
#include <list>
#include <map>

using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        vector<int> dp(nums.size() + 1, 0);
        dp[0] = 0;
        dp[1] = nums[0];

        for (int i = 2; i <= nums.size(); i++) {
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[i - 1]);
        }

        return dp[nums.size()];
    }
};

int main(int argc, char *argv[]) {
    Solution solution;
    vector<int> nums;

    // Input: nums = [1,2,3,1]
    // Output: 4
    nums = {1, 2, 3, 1};
    cout << solution.rob(nums) << endl;

    // Input: nums = [2,7,9,3,1]
    // Output: 12
    nums = {2, 7, 9, 3, 1};
    cout << solution.rob(nums) << endl;

    return 0;
}

File: py/0198_House_Robber.py
import sys
import heapq
from typing import *
from collections import OrderedDict
from collections import deque


class Solution:
    '''
    def helper(self, nums, i, memo):
        if i < 0:
            return 0

        if i in memo:
            return memo[i]

        memo[i] = max(self.helper(nums, i - 2, memo) + nums[i], self.helper(nums, i - 1, memo))

        return memo[i]

    def rob(self, nums: List[int]) -> int:
        memo = {}
        return self.helper(nums, len(nums) - 1, memo)
    '''

    def rob(self, nums: List[int]) -> int:
        dp = [0] * len(nums)
        if len(nums) == 0: return 0
        if len(nums) == 1: return nums[0]
        if len(nums) == 2: return max(nums)
        dp = [0] * len(nums)
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])

        for i in range(2, len(nums)):
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[i])

        return dp[-1]



if __name__ == "__main__":
    solution = Solution()

    nums = [1, 2, 3, 1]
    print(solution.rob(nums))

    nums = [2, 7, 9, 3, 1]
    print(solution.rob(nums))

    sys.exit(0)
